06-r-and-sql.Rmd

```diff
@@ -0,0 +1,144 @@
+FROM species a
+JOIN surveys b
+ON a.species_id = b.species_id) c
+JOIN plots d
+ON c.plot = d.plot_id
+GROUP BY d.plot_type,c.genus"
+
+result <- dbGetQuery(conn,q)
+head(result)
+
+```
+
+All we've done so far is execute the same sorts of queries that can easily be made with a GUI.  Now let's try leveraging some of the power scripted queries.  Imagine you want to know how many rodents were found every other year.  To get this we'll get the range of dates from the database, sequence them by two and make new queries.
+
+```{r Get years of data}
+yearRange <- dbGetQuery(conn,"SELECT min(year),max(year) from surveys")
+years <- seq(yearRange[,1],yearRange[,2],by=2)
+```
+
+Next we'll build our query string using the `paste()` function.
+
+```{r creating query}
+q <- paste("
+SELECT a.year,b.taxa,count(*) as count
+FROM surveys a
+JOIN species b
+ON a.species_id = b.species_id
+AND b.taxa = 'Rodent'
+AND a.year in (",
+paste(years,collapse=",")
+,")
+GROUP BY a.year, b.taxa",
+sep = "" )
+rCount <- dbGetQuery(conn,q)
+head(rCount)
+```
+
+With the nested paste commands we were able to construct a query programatically, without having to type out all the years.  This could also be done with a for loop, especially if the query to be constructed is more complicated.
+
+Building your workflow
+------
+
+Up until now we been working with an existing data. However we can use R as way to build up databases from existing flat files.  We'll use data we generate, but you could easily do this with flat files that you read into R.  Here we'll simulate a simple file of 'sites' where you've measured snout vent lengths (svn) of lizards.
+
+```{r Generate a dataframe}
+svn_lengths <- cbind(data.frame(sample(letters[1:5],500,replace=T)),data.frame(rnorm(500,15,6)))
+colnames(svn_lengths) <- c("site","length")
+```
+
+We have some data now. Next we'll need to create a database and add it as a table.
+
+```{r create database}
+myDB <- "~/wkspace/test.db"
+myConn <- dbConnect(drv = SQLite(), dbname= myDB)
+dbListTables(myConn)
+```
+By opening up a connection, you've created a new database.  If you list the tables you'll see that the database is empty. Next we'll add the dataframe we just created to the database.
+
+```{r add data}
+dbWriteTable(myConn,"svn",svn_lengths)
+dbListTables(myConn)
+dbGetQuery(myConn,"SELECT * from svn limit 10")
+```
+  
+If you check the location of your database you'll see that data is automatically being written to disk. Not only does R and RSQLite provide easy ways to query existing databases, it also allows you to easily create your own databases from raw data.  Using this method, you can not only store raw data but create linked tables of derived metrics or data from external sources.
+
+>  **Challenge**
+>
+> Add a new table to the existing database.  Perhaps it's data about the five different sites, or any other table you'd like. Query your database from SQLite manager just to verify that you're adding data.
```

Comment:
I might make this last challenge two parts? e.g. 1) add a second table to the database (and maybe verify via SQLite), then 2) query something that requires a join and then plot or calculate some statistic from that joined data. Just a thought--it might help to bring it back full circle to producing analysis products. Or is that too much.